common/hashing_test.cpp

```diff
@@ -288,53 +350,72 @@ auto FindBitRangeCollisions(llvm::ArrayRef<HashedValue<T>> hashes)
     bits_and_indices.push_back(
         {static_cast<uint32_t>(hash_bits), bits_and_indices.size()});
   }
+
+  // Now we sort by the extracted bit sequence so we can efficiently scan for
+  // colliding bit patterns.
   std::sort(
       bits_and_indices.begin(), bits_and_indices.end(),
       [](const auto& lhs, const auto& rhs) { return lhs.first < rhs.first; });
 
+  // Scan the sorted bit sequences we've extracted looking for collisions. We
+  // count the total collisions, but we also track the number of individual
+  // inputs that collide with each specific bit pattern.
   uint32_t prev_hash_bits = bits_and_indices[0].first;
   int prev_index = bits_and_indices[0].second;
   bool in_collision = false;
   int total = 0;
-  int distinct_collisions = 0;
   for (const auto& [hash_bits, hash_index] :
        llvm::ArrayRef(bits_and_indices).slice(1)) {
+    // Check if we've found a new hash (and thus a new value), reset everything.
     if (hash_bits != prev_hash_bits) {
+      CARBON_CHECK(hashes[prev_index].hash != hashes[hash_index].hash);
+      CARBON_CHECK(hashes[prev_index].v != hashes[hash_index].v);
       prev_hash_bits = hash_bits;
       prev_index = hash_index;
       in_collision = false;
       continue;
     }
 
+    // Otherwise, we have a colliding bit sequence.
     ++total;
+
+    // If we've already created a collision count to track this, just increment
+    // it and map this hash to it.
     if (in_collision) {
       ++collision_counts.back();
       collision_map[hash_index] = collision_counts.size() - 1;
       continue;
     }
+
+    // If this is a new collision, create a dedicated count to track it and
+    // begin counting.
     in_collision = true;
     collision_map[hash_index] = collision_counts.size();
```

Comment:
Without this, the number of items with 0 collisions will be over-counted in the median calculation.
```suggestion
    collision_map[prev_index] = collision_counts.size();
    collision_map[hash_index] = collision_counts.size();
```